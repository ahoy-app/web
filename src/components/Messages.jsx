import React, { useEffect, useRef } from 'react'
import PropTypes from 'prop-types'

import MessageRow from './MessageRow'

const styles = {
  height: '100%',
  width: '100%',

  overflowY: 'scroll',

  padding: '5px 10px 5px 10px',
  boxSizing: 'border-box',

  display: 'flex',
  flexDirection: 'column',
  alignItems: 'flex-start',
}

const Messages = ({ messages, fetchOldMessages }) => {
  const anchorEl = useRef()
  const boxEl = useRef()

  useEffect(() => {
    anchorEl.current.scrollIntoView({ behavior: 'smooth' })
  }, messages.length)

  const _onScroll = () => {
    if (boxEl.current.scrollTop === 0) {
      console.log(fetchOldMessages())
    }
  }

  return (
    <div style={styles} ref={boxEl} onScroll={_onScroll}>
      {messages.map(message => (
        <MessageRow message={message} key={message.id} />
      ))}
      <div ref={anchorEl} />
    </div>
  )
}

Messages.propTypes = {
  messages: PropTypes.array.isRequired,
  fetchOldMessages: PropTypes.func.isRequired,
}
// export default Radium(Messages)
export default Messages
