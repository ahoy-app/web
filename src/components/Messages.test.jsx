import React from 'react'
import { mount } from 'enzyme'

import Messages from './Messages'

const messages = [
  {
    id: '12we4tq34t9guw1344tt',
    from: 'Barney',
    content: "It's gona be legend.. Whait for it!",
  },
  {
    id: 'q384tp1349fgh139p48g',
    from: '-',
    content: 'See you later',
  },
  {
    id: '823498t1394gt13t1448',
    from: 'Barney',
    content: '..dary! Legendary, Ted!',
  },
]

describe('Messages', () => {
  beforeAll(() => {
    if (!HTMLElement.prototype.scrollIntoView) {
      HTMLElement.prototype.scrollIntoView = jest.fn()
    }
  })

  it('should render correctly', () => {
    const component = mount(
      <Messages messages={messages} fetchOldMessages={() => {}} />
    )
    expect(component).toMatchSnapshot()
  })

  it('should call fetchOldMessages when scrolled up', () => {
    const fetchFn = jest.fn()

    const component = mount(
      <Messages messages={messages} fetchOldMessages={fetchFn} />
    )

    component.simulate('scroll')

    expect(fetchFn).toHaveBeenCalled()
  })
})
