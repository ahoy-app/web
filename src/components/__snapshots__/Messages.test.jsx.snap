// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Messages should render correctly 1`] = `
<Messages
  fetchOldMessages={[Function]}
  messages={
    Array [
      Object {
        "content": "It's gona be legend.. Whait for it!",
        "from": "Barney",
        "id": "12we4tq34t9guw1344tt",
      },
      Object {
        "content": "See you later",
        "from": "-",
        "id": "q384tp1349fgh139p48g",
      },
      Object {
        "content": "..dary! Legendary, Ted!",
        "from": "Barney",
        "id": "823498t1394gt13t1448",
      },
    ]
  }
>
  <div
    onScroll={[Function]}
    style={
      Object {
        "alignItems": "flex-start",
        "boxSizing": "border-box",
        "display": "flex",
        "flexDirection": "column",
        "height": "100%",
        "overflowY": "scroll",
        "padding": "5px 10px 5px 10px",
        "width": "100%",
      }
    }
  >
    <MessageRow
      key="12we4tq34t9guw1344tt"
      message={
        Object {
          "content": "It's gona be legend.. Whait for it!",
          "from": "Barney",
          "id": "12we4tq34t9guw1344tt",
        }
      }
    >
      <div
        data-radium={true}
        style={
          Object {
            "alignSelf": "flex-start",
            "boxSizing": "border-box",
            "padding": "5px 0px 5px 0px",
          }
        }
      >
        <Message
          message={
            Object {
              "content": "It's gona be legend.. Whait for it!",
              "from": "Barney",
              "id": "12we4tq34t9guw1344tt",
            }
          }
        >
          <div
            data-radium={true}
            style={
              Object {
                "background": "linear-gradient(135deg, #fff 0%, #e8e8e8 100%)",
                "borderRadius": "10px",
                "boxSizing": "border-box",
                "display": "flex",
                "filter": "drop-shadow(2px 5px 2px #0003)",
                "flexDirection": "column",
                "minHeight": "40px",
                "padding": "5px 10px 5px 10px",
                "width": "500px",
              }
            }
          >
            <span
              key=".0"
            >
              Barney
            </span>
            <span
              key=".1"
            >
              It's gona be legend.. Whait for it!
            </span>
          </div>
        </Message>
      </div>
    </MessageRow>
    <MessageRow
      key="q384tp1349fgh139p48g"
      message={
        Object {
          "content": "See you later",
          "from": "-",
          "id": "q384tp1349fgh139p48g",
        }
      }
    >
      <div
        data-radium={true}
        style={
          Object {
            "alignSelf": "flex-end",
            "boxSizing": "border-box",
            "padding": "5px 0px 5px 0px",
          }
        }
      >
        <Message
          message={
            Object {
              "content": "See you later",
              "from": "-",
              "id": "q384tp1349fgh139p48g",
            }
          }
        >
          <div
            data-radium={true}
            style={
              Object {
                "background": "linear-gradient(135deg, #fff 0%, #e8e8e8 100%)",
                "borderRadius": "10px",
                "boxSizing": "border-box",
                "display": "flex",
                "filter": "drop-shadow(2px 5px 2px #0003)",
                "flexDirection": "column",
                "minHeight": "40px",
                "padding": "5px 10px 5px 10px",
                "width": "500px",
              }
            }
          >
            <span
              key=".0"
            >
              -
            </span>
            <span
              key=".1"
            >
              See you later
            </span>
          </div>
        </Message>
      </div>
    </MessageRow>
    <MessageRow
      key="823498t1394gt13t1448"
      message={
        Object {
          "content": "..dary! Legendary, Ted!",
          "from": "Barney",
          "id": "823498t1394gt13t1448",
        }
      }
    >
      <div
        data-radium={true}
        style={
          Object {
            "alignSelf": "flex-start",
            "boxSizing": "border-box",
            "padding": "5px 0px 5px 0px",
          }
        }
      >
        <Message
          message={
            Object {
              "content": "..dary! Legendary, Ted!",
              "from": "Barney",
              "id": "823498t1394gt13t1448",
            }
          }
        >
          <div
            data-radium={true}
            style={
              Object {
                "background": "linear-gradient(135deg, #fff 0%, #e8e8e8 100%)",
                "borderRadius": "10px",
                "boxSizing": "border-box",
                "display": "flex",
                "filter": "drop-shadow(2px 5px 2px #0003)",
                "flexDirection": "column",
                "minHeight": "40px",
                "padding": "5px 10px 5px 10px",
                "width": "500px",
              }
            }
          >
            <span
              key=".0"
            >
              Barney
            </span>
            <span
              key=".1"
            >
              ..dary! Legendary, Ted!
            </span>
          </div>
        </Message>
      </div>
    </MessageRow>
    <div />
  </div>
</Messages>
`;
